commons/src/main/java/org/infinispan/commons/jdkspecific/CallerId.java

```diff
@@ -0,0 +1,40 @@
+package org.infinispan.commons.jdkspecific;
+
+import sun.reflect.Reflection;
+
+public class CallerId {
+   private static final boolean hasGetCallerClass;
+   private static final int callerOffset;
+   private static final LocalSecurityManager SECURITY_MANAGER;
+
+   static {
+      boolean result = false;
+      int offset = 0;
+      try {
+         result = Reflection.getCallerClass(1) == CallerId.class || Reflection.getCallerClass(2) == CallerId.class;
+         offset = Reflection.getCallerClass(1) == Reflection.class ? 2 : 1;
+      } catch (Throwable ignored) {
+      }
+      hasGetCallerClass = result;
+      callerOffset = offset;
+      if (!hasGetCallerClass) {
+         SECURITY_MANAGER = new LocalSecurityManager();
+      } else {
+         SECURITY_MANAGER = null;
+      }
+   }
+
+   private static class LocalSecurityManager extends SecurityManager {
+      public Class<?>[] getClasses() {
+         return this.getClassContext();
+      }
+   }
+
+   public static Class<?> getCallerClass(int n) {
+      if (hasGetCallerClass) {
+         return Reflection.getCallerClass(n + callerOffset);
+      } else {
+         return SECURITY_MANAGER.getClasses()[n + callerOffset];
```

Comment:
I _borrowed_ this code from the WildFly Security Manager. I guess we can remove the + callerOffset

Reply:
I'm not sure that's ok, I think we want to add 1 even with `SECURITY_MANAGER.getClasses()`. 
Do we have any tests that force the security manager path?

Reply:
Yeah, it's good enough.

I'm not 100% satisfied because I have no idea when/if we could have `hasGetCallerClass == false` without `Reflection.getCallerClass(1)` throwing an exception, but maybe it's best not to dwell too much on such questions :)